File: packages/ytl-linux-digabi2-examnet/digabi2-examnet-bouncer/src/digabi2_examnet_bouncer/bouncer.py
from collections.abc import Callable, Iterable, Sequence
import functools
from typing import final
from urllib.parse import urljoin
from wsgiref.types import StartResponse, WSGIEnvironment

from werkzeug import Request, Response
from werkzeug.exceptions import HTTPException, NotFound
from werkzeug.routing import Map, Rule

from .config import Config


NCSI_RESPONSES = {
    "connecttest.txt": "Microsoft Connect Test",
    "ncsi.txt": "Microsoft NCSI",
}


def format_hosts(port: int, hostnames: Sequence[str]):
    def acknowledge_trailing_dot(it: Iterable[str]):
        for hostname in it:
            yield hostname
            yield hostname.removesuffix(".")

    def insert_port(it: Iterable[str]):
        for hostname in it:
            yield hostname if port == 80 else  f"{hostname}:{port}"

    return insert_port(acknowledge_trailing_dot(hostnames))


@final
class Bouncer:
    def __init__(self, config: Config, friendly_hostname: str):
        self.config = config
        port = config.bouncer_port

        ncsi_template = functools.partial(Rule, "/<name>", endpoint=self.ncsi, methods=["GET"])
        ncsi_hosts = set(format_hosts(port, self.config.ncsi_hosts))
        ncsi_rules = [ncsi_template(host=host) for host in ncsi_hosts]

        mdns_routes = [("/", ["GET"]), ("/ktp/hello", ["POST"])]
        mdns_hosts = set(format_hosts(port, [friendly_hostname]))
        mdns_rules = [
            Rule(route, endpoint=self.redirect, methods=methods, host=host)
            for route, methods in mdns_routes
            for host in mdns_hosts
        ]

        self.url_map = Map([*ncsi_rules, *mdns_rules], host_matching=True)

    def dispatch_request(self, request: Request):
        adapter = self.url_map.bind_to_environ(request.environ)
        try:
            rule, values = adapter.match(return_rule=True)
            endpoint: Callable[[Request], Response] = rule.endpoint  # pyright: ignore[reportAny]
            return endpoint(request, **values)
        except HTTPException as exc:
            return exc

    def __call__(self, environ: WSGIEnvironment, start_response: StartResponse):
        request = Request(environ)
        response = self.dispatch_request(request)
        return response(environ, start_response)

    def ncsi(self, _request: Request, *, name: str) -> Response:
        try:
            response = NCSI_RESPONSES[name]
        except KeyError:
            raise NotFound()
        return Response(response, 200, content_type="text/plain")

    def redirect(self, request: Request):
        headers = {
            "Access-Control-Allow-Origin": "*",
            "Location": urljoin(self.config.canonical_url, request.full_path),
        }
        return Response(status=307, headers=headers)

File: packages/ytl-linux-digabi2-examnet/bouncer2/src/bouncer.ts
import { match } from '@gabriel/ts-pattern'
import { Config } from './config.ts'

export function bouncerApp(config: Config): (req: Request) => Response {
  const { ncsiHostnames, friendlyName } = config
  const friendlyHostname = `${friendlyName}.${config.searchDomain}`

  return function bouncerHandler(req: Request): Response {
    const url = new URL(req.url)
    return match(url.hostname)
      .when(
        x => x === friendlyHostname || x === friendlyName,
        () => redirectHandler(req)
      )
      .when(
        x => ncsiHostnames.includes(x),
        () => ncsiHandler(req)
      )
      .otherwise(() => new Response(null, { status: 404 }))
  }

  function ncsiHandler(req: Request): Response {
    const url = new URL(req.url)
    return match(url.pathname)
      .with('/connecttest.txt', () => new Response('Microsoft Connect Test'))
      .with('/ncsi.txt', () => new Response('Microsoft NCSI'))
      .otherwise(() => new Response(null, { status: 404 }))
  }

  function redirectHandler(req: Request): Response {
    const url = new URL(req.url)
    return match([req.method, url.pathname.replace(/\/$/, '')])
      .with(['GET', '/valvoja'], () => {
        const targetURL = new URL(url)
        targetURL.protocol = 'https'
        targetURL.host = config.canonicalHostname
        targetURL.pathname = '/'
        return redirect(targetURL)
      })
      .with(['GET', '/koe'], () => {
        const targetURL = new URL(url)
        targetURL.protocol = 'https'
        targetURL.hostname = config.canonicalHostname
        targetURL.port = '8010'
        targetURL.pathname = '/'
        return redirect(targetURL)
      })
      .with(['POST', '/ktp/hello'], () => {
        const targetURL = new URL(url)
        targetURL.protocol = 'https'
        targetURL.hostname = config.canonicalHostname
        targetURL.port = '8010'
        return redirect(targetURL)
      })
      .otherwise(() => new Response(null, { status: 404 }))
  }
}

function redirect(targetURL: URL) {
  return new Response(null, {
    status: 307,
    headers: {
      'Access-Control-Allow-Origin': '*',
      Location: targetURL.toString()
    }
  })
}
